File: leetcode/3041.修改数组后最大化数组中的连续元素数目.py
class Solution:
    def maxSelectedElements(self, nums: List[int]) -> int:
        l = min(nums)
        r = max(nums)
        dic =defaultdict(int)
        nums.sort()
        for num in nums:
            dic[num]+=1
        to = defaultdict(int)
        for k,d in dic.items():
            if k-1 in dic:
                to[k]=to[k-1]+1
            else:
                to[k]=1
        mxn = 1
        ptr = l-1
        for i in range(l,r+3):
            if dic[i] == 0 and dic[i-1]<2:
                mxn = max(i-1-ptr+to[ptr-1],mxn)
                ptr = i
                continue
            if dic[i-1]>=2:
                dic[i]+=1
        return mxn

File: leetcode/3041.修改数组后最大化数组中的连续元素数目.go
func maxSelectedElements(nums []int) int {
    // dp[i]表示以i为结尾的最长长度
    // 
    dp := make(map[int]int)
    sort.Ints(nums)
    mx := 0
    for _, num := range nums {
        dp[num + 1] = max(dp[num + 1], dp[num] + 1)
        dp[num] = max(dp[num], dp[num - 1] + 1)
        mx = max(max(mx, dp[num]), dp[num + 1])
    }
    return mx
}